lrauv_ignition_plugins/src/ScienceSensorsSystem.cc

```diff
@@ -934,23 +1046,166 @@ float ScienceSensorsSystemPrivate::TrilinearInterpolate(
     return std::numeric_limits<float>::quiet_NaN();
   }
 
-  // Find closest neighbor
-  Eigen::MatrixXf::Index minR, minC;
+  // A rectangular prism can be represented by two pairs of 3D coordinates,
+  // (x0, y0, z0) and (x1, y1, z1), which are diagonal vertices on the prism.
+  // Extract 2 diagonal vertices to represent the rectangular prism, assuming
+  // the points are corners of a prism (assumption checked after extracting).
+  // Corner of minimum x y z
+  auto v000 = Eigen::Vector3f(
+    _xyzs.col(0).minCoeff(),
+    _xyzs.col(1).minCoeff(),
+    _xyzs.col(2).minCoeff());
+  // Corner of maximum x y z
+  auto v111 = Eigen::Vector3f(
+    _xyzs.col(0).maxCoeff(),
+    _xyzs.col(1).maxCoeff(),
+    _xyzs.col(2).maxCoeff());
+  // 6 remaining vertices of prism
+  Eigen::Vector3f v001, v010, v011, v100, v101, v110;
+
+  igndbg << "Trilinear interpolation min vert v000: "
+     << v000(0) << ", " << v000(1) << ", " << v000(2) << std::endl;
+  igndbg << "Trilinear interpolation max vert v111: "
+     << v111(0) << ", " << v111(1) << ", " << v111(2) << std::endl;
+
+  // Data values at the vertices
+  // Define explicitly for readability in interpolation equations later
+  float d000, d001, d010, d011, d100, d101, d110, d111;
+
+  // Tolerance below which to consider two values equal
+  const double TOLERANCE = 1e-6;
+
+  // Find ordering of corners in prism, assign data to ordered corners.
+  // Each vertex, if really on the corner of a rectangular prism, must share
```

Comment:
It's not necessary to have a purely cubic solution. We can do bilinear interpolation between the 4 closest points in each slice as long as they're surrounding the point. And then it's just linear interpolation vertically. 

We can then relax the pure cube which is axis alligned requirement in the algorithm. An example where this might be useful would be if there's different data densities at different levels. Such as potentially much higher at the surface where sampling is easier.

Reply:
So, I think it is better to add that as an additional enhancement, rather than replacing what we already have, since the one we have is a stricter algorithm and might give different results. Note that the hybrid bary-linear method is already a relaxation of the trilinear method.

Like it's been verbally communicated at the end of last year, there's no perfect solution to interpolation. We have 3 alternatives now (barycentric, linear, bary-linear), and that should give whoever using this code choices to play with, when there is time to really evaluate and compare, which interpolation method works better for which application. We don't have any data to choose one over another, so we're just keeping different alternatives around.